### PR TITLE
Refactoring the translation of execve() and fixing the handling of shebang

### PR DESCRIPTION
**Pull request by KB5201314**
This PR contains two main parts:

- Refactoring the translation of execve()
  - The old execve() translation function does not implement recursive parsing of the interpreter, which means that the interpreter of a script file can not be a script file, resulting in a limitation. This PR fixed that problem.
  - The previous process of loading executable files is less extensible. For this reason, I split the load function into `load_elf()` and `load_script()`.

- Fixing the handling of shebang
  - As mentioned in https://github.com/proot-me/proot-rs/issues/8#issuecomment-807059898, the argument in the shebang line of script file are not handled correctly. This PR rewrites the parsing of script files.
  - Added integration test for loading script file.

### CODE REVIEW

File: src/kernel/execve/binfmt/elf.rs [hunk 1]
```diff
@@ -231,6 +236,30 @@ impl ElfHeader {
...
+pub(super) fn load_elf(fs: &FileSystem, parameters: &mut ExecveParameters) -> Result<LoadResult> {
+    // parse LoadInfo from the binary file to be executed
+    let mut load_info = LoadInfo::from(fs, &parameters.host_path)
+        .with_context(|| format!("Failed to parse elf file: {:?}", parameters.host_path))?;
+
+    load_info.raw_path = Some(parameters.raw_guest_path.clone());
+    load_info.user_path = Some(parameters.canonical_guest_path.clone());
+    load_info.host_path = Some(parameters.host_path.clone());
+
+    // An interpreter should not depend on another interpreter
```
Comment (oxr463) [inline src/kernel/execve/binfmt/elf.rs:249]: Any reason why you have this comment here? Do you have any references? Or did this happen from a practical example? e.g., running a command

File: tests/execve/test.bats [hunk 2]
```diff
@@ -20,4 +20,76 @@ load ../helper
...
+    # Test optional argument
+    echo '#!/bin/echo "123"' > ./script4.sh
+    chmod +x ./script4.sh
+    [ "$(./script4.sh 2>&1)" = '"123" ./script4.sh' ]
+
+    echo '#!/bin/echo "123      "' > ./script5.sh
+    chmod +x ./script5.sh
+    [ "$(./script5.sh 2>&1)" = '"123      " ./script5.sh' ]
+
+    # Blank-space in the middle of opt argument is reserved. But the ones at the beginning and the end are stripped
```
Comment (oxr463) [inline tests/execve/test.bats:59]: reserved? or _preserved_?

**Review by oxr463 — APPROVED** [review]

File: src/kernel/execve/binfmt/elf.rs (hunk 1 above)
Comment (KB5201314) [reply src/kernel/execve/binfmt/elf.rs:249]: > Any reason why you have this comment here? Do you have any references? Or did this happen from a practical example? e.g., running a command

This comment comes from the C version of proot.
```C
		 /* An ELF interpreter is supposed to be
		 * standalone.  */
```
https://github.com/proot-me/proot/blob/20d619d907e7c80dc33b884a93d0645c7ea96cc2/src/execve/enter.c#L664-L665

Although there is no documentation that explicitly mentions that the ELF interpreter should be statically linked (i.e. no PT_INTERP segment). But from [this article](https://lwn.net/Articles/631631/) on how does kernel load ELF files, it appears that when loading the ELF interpreter, the handler does not handle segments of type `PT_INTERP`, only `PT_LOAD`. 

> After the program code has been loaded into memory as described previously, the ELF handler also loads the ELF interpreter program into memory with load_elf_interp(). This process is similar to the process of loading the original program: the code checks the format information in the ELF header, reads in the ELF program header, **maps all of the PT_LOAD segments from the file into the new program's memory**, and leaves room for the interpreter's BSS segment.

The kernel source code also confirms this 
https://elixir.bootlin.com/linux/v3.18/source/fs/binfmt_elf.c#L452

Also, realistically, as the most common ELF interpreter, `/lib64/ld-linux-x86-64.so.2` is indeed statically linked.
image

File: tests/execve/test.bats (hunk 2 above)
Comment (KB5201314) [reply tests/execve/test.bats:59]: > reserved? or _preserved_?

Maybe "preserved" would be better 😂

Comment (KB5201314) [reply tests/execve/test.bats]: Already changed in the new commit

**Pull request merged**